docs: add redirects for docs IA update

bpmutter opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Redirects for docs site as part of the information architecture overhaul.

Should be merged together with https://github.com/eslint/eslint/pull/16665

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

Added redirects for new docs information architecture.

#### Related Issues

https://github.com/eslint/eslint/pull/16665

#### Is there anything you'd like reviewers to focus on?

Note: created some helper scripts in this repo - https://github.com/bpmutter/eslint-redirects

This can (but doesn't have to be) used to make any bulk changes to the additions here. 

<!-- markdownlint-disable-file MD004 -->

bpmutter commented on src/static/redirects.njk:
```diff
@@ -51,3 +51,42 @@ eleventyExcludeFromCollections: true
+/docs/latest/maintainer-guide/issues/                        /docs/latest/maintain/manage-issues/ 301!
+/docs/latest/maintainer-guide/pullrequests/                  /docs/latest/maintain/review-pull-requests/ 301!
+/docs/latest/maintainer-guide/releases/                      /docs/latest/maintain/manage-releases/ 301!
+/docs/latest/maintainer-guide/governance/                    /docs/latest/contribute/governance/ 301!
```
```suggestion
/docs/latest/user-guide/configuring/*                    /docs/latest/use/contribute/:splat/ 301!
```

mdjermanovic reviewed (commented):
This doesn't seem to work. For example, https://deploy-preview-388--new-eslint.netlify.app/docs/latest/user-guide/ does not redirect to https://deploy-preview-388--new-eslint.netlify.app/docs/latest/use/

Perhaps the new redirect rules should be before this one so that the new rules match first:

https://github.com/eslint/eslint.org/blob/52f9eb583db914f2783f778f810103a08d01ad63/src/static/redirects.njk#L34-L35

https://docs.netlify.com/routing/redirects/#rule-processing-order

bpmutter commented:
> This doesn't seem to work. For example, https://deploy-preview-388--new-eslint.netlify.app/docs/latest/user-guide/ does not redirect to https://deploy-preview-388--new-eslint.netlify.app/docs/latest/use/
> 
> Perhaps the new redirect rules should be before this one so that the new rules match first:
> 
> https://github.com/eslint/eslint.org/blob/52f9eb583db914f2783f778f810103a08d01ad63/src/static/redirects.njk#L34-L35
> 
> https://docs.netlify.com/routing/redirects/#rule-processing-order

didn't realize that there was a staging site that redirects would work on here. thx for call out. will play around w that for testing

nzakas commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/maintainer-guide/pullrequests/                  /docs/latest/maintain/review-pull-requests/ 301!
+/docs/latest/maintainer-guide/releases/                      /docs/latest/maintain/manage-releases/ 301!
+/docs/latest/maintainer-guide/governance/                    /docs/latest/contribute/governance/ 301!
+/docs/latest/user-guide/configuring/*                        /docs/latest/use/contribute/:splat/ 301!
+/docs/latest/user-guide/*                                    /docs/latest/use/:splat/ 301!
+/docs/latest/developer-guide/*                               /docs/latest/extend/:splat/ 301!
+/docs/latest/maintainer-guide/*                              /docs/latest/maintain/:splat/ 301!
```
When using `:splat`, leave off the trailing slash, otherwise we'll get two trailing slashes on the redirect in the browser URL bar.

```suggestion
/docs/latest/user-guide/configuring/*                        /docs/latest/use/contribute/:splat 301!
/docs/latest/user-guide/*                                    /docs/latest/use/:splat 301!
/docs/latest/developer-guide/*                               /docs/latest/extend/:splat 301!
/docs/latest/maintainer-guide/*                              /docs/latest/maintain/:splat 301!
```

nzakas reviewed (dismissed):
I did a spot-check of a few links and they appear to redirect correctly now. Just a bit of cleanup with regards to splatted URLs.

bpmutter commented:
> > This doesn't seem to work. For example, https://deploy-preview-388--new-eslint.netlify.app/docs/latest/user-guide/ does not redirect to https://deploy-preview-388--new-eslint.netlify.app/docs/latest/use/
> > Perhaps the new redirect rules should be before this one so that the new rules match first:
> > https://github.com/eslint/eslint.org/blob/52f9eb583db914f2783f778f810103a08d01ad63/src/static/redirects.njk#L34-L35
> > 
> > https://docs.netlify.com/routing/redirects/#rule-processing-order
> 
> didn't realize that there was a staging site that redirects would work on here. thx for call out. will play around w that for testing

ok, implemented the suggested change, and it's working as expected. using example above, now <https://deploy-preview-388--new-eslint.netlify.app/docs/latest/user-guide/> redirects to <https://deploy-preview-388--new-eslint.netlify.app/docs/latest/use/>

it also captures the splat patterns such as <https://deploy-preview-388--new-eslint.netlify.app/docs/latest/user-guide/page-does-not-exist> to <<https://deploy-preview-388--new-eslint.netlify.app/docs/latest/use/page-does-not-exist>

---

with that being said, there's a lot of conditional logic going on in this file that i don't fully understand, so it'd be great if this could be reviewed taking that into account.

nzakas commented:
The conditional logic is primarily for the various translation sites, only one of which (zh-hans.eslint.org) has its own docs.

Although that does remind me: when we are renaming files, we need to make sure to do it for the zh-hans docs as well or else everything will break.

bpmutter commented:
> The conditional logic is primarily for the various translation sites, only one of which (zh-hans.eslint.org) has its own docs.
> 
> Although that does remind me: when we are renaming files, we need to make sure to do it for the zh-hans docs as well or else everything will break.

ok, how do you think is best to proceed with this? 

i don't know chinese so am probably not too useful with that.

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/user-guide/migrating-to-8.0.0/                  /docs/latest/use/migrate-to-8.0.0/ 301!
+/docs/latest/developer-guide/architecture/                   /docs/latest/contribute/architecture/ 301!
+/docs/latest/developer-guide/source-code/                    /docs/latest/contribute/development-environment/ 301!
+/docs/latest/developer-guide/development-environment/        /docs/latest/contribute/development-environment/ 301!
+/docs/latest/developer-guide/unit-tests/                     /docs/latest/contribute/tests/ 301!
```
```suggestion
/docs/latest/developer-guide/development-environment         /docs/latest/contribute/development-environment 301!
/docs/latest/developer-guide/unit-tests                      /docs/latest/contribute/tests 301!
```

To redirect to canonical URLs.

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/developer-guide/source-code/                    /docs/latest/contribute/development-environment/ 301!
+/docs/latest/developer-guide/development-environment/        /docs/latest/contribute/development-environment/ 301!
+/docs/latest/developer-guide/unit-tests/                     /docs/latest/contribute/tests/ 301!
+/docs/latest/developer-guide/working-with-rules/             /docs/latest/extend/custom-rules/ 301!
+/docs/latest/developer-guide/working-with-plugins/           /docs/latest/extend/plugins/ 301!
+/docs/latest/developer-guide/working-with-custom-formatters/ /docs/latest/extend/custom-formatters/ 301!
+/docs/latest/developer-guide/working-with-custom-parsers/    /docs/latest/extend/custom-parsers/ 301!
```
```suggestion
/docs/latest/developer-guide/working-with-rules              /docs/latest/extend/custom-rules 301!
/docs/latest/developer-guide/working-with-plugins            /docs/latest/extend/plugins 301!
/docs/latest/developer-guide/working-with-custom-formatters  /docs/latest/extend/custom-formatters 301!
/docs/latest/developer-guide/working-with-custom-parsers     /docs/latest/extend/custom-parsers 301!
```

To redirect to canonical URLs.

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/developer-guide/working-with-plugins/           /docs/latest/extend/plugins/ 301!
+/docs/latest/developer-guide/working-with-custom-formatters/ /docs/latest/extend/custom-formatters/ 301!
+/docs/latest/developer-guide/working-with-custom-parsers/    /docs/latest/extend/custom-parsers/ 301!
+/docs/latest/developer-guide/code-path-analsis/              /docs/latest/extend/code-path-analsis/ 301!
```
```suggestion
/docs/latest/developer-guide/code-path-analysis              /docs/latest/extend/code-path-analysis 301!
```

Same + typos.

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/developer-guide/working-with-custom-formatters/ /docs/latest/extend/custom-formatters/ 301!
+/docs/latest/developer-guide/working-with-custom-parsers/    /docs/latest/extend/custom-parsers/ 301!
+/docs/latest/developer-guide/code-path-analsis/              /docs/latest/extend/code-path-analsis/ 301!
+/docs/latest/developer-guide/nodejs-api/                     /docs/latest/integrate/nodejs-api/ 301!
```
```suggestion
/docs/latest/developer-guide/nodejs-api                      /docs/latest/integrate/nodejs-api 301!
```

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/developer-guide/code-path-analsis/              /docs/latest/extend/code-path-analsis/ 301!
+/docs/latest/developer-guide/nodejs-api/                     /docs/latest/integrate/nodejs-api/ 301!
+/docs/latest/developer-guide/contributing/                   /docs/latest/contribute/ 301!
+/docs/latest/developer-guide/contributing/reporting-bugs/    /docs/latest/contribute/report-bugs/ 301!
+/docs/latest/developer-guide/contributing/new-rules/         /docs/latest/contribute/propose-new-rule/ 301!
+/docs/latest/developer-guide/contributing/rule-changes/      /docs/latest/contribute/propose-rule-change/ 301!
+/docs/latest/developer-guide/contributing/changes/           /docs/latest/contribute/request-change/ 301!
+/docs/latest/developer-guide/contributing/working-on-issues/ /docs/latest/contribute/work-on-issue/ 301!
+/docs/latest/developer-guide/contributing/pull-requests/     /docs/latest/contribute/pull-requests/ 301!
```
```suggestion
/docs/latest/developer-guide/contributing/reporting-bugs    /docs/latest/contribute/report-bugs 301!
/docs/latest/developer-guide/contributing/new-rules          /docs/latest/contribute/propose-new-rule 301!
/docs/latest/developer-guide/contributing/rule-changes       /docs/latest/contribute/propose-rule-change 301!
/docs/latest/developer-guide/contributing/changes            /docs/latest/contribute/request-change 301!
/docs/latest/developer-guide/contributing/working-on-issues  /docs/latest/contribute/work-on-issue 301!
/docs/latest/developer-guide/contributing/pull-requests      /docs/latest/contribute/pull-requests 301!
```

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/developer-guide/contributing/working-on-issues/ /docs/latest/contribute/work-on-issue/ 301!
+/docs/latest/developer-guide/contributing/pull-requests/     /docs/latest/contribute/pull-requests/ 301!
+/docs/latest/maintainer-guide/                               /docs/latest/maintain/ 301!
+/docs/latest/maintainer-guide/issues/                        /docs/latest/maintain/manage-issues/ 301!
+/docs/latest/maintainer-guide/pullrequests/                  /docs/latest/maintain/review-pull-requests/ 301!
+/docs/latest/maintainer-guide/releases/                      /docs/latest/maintain/manage-releases/ 301!
+/docs/latest/maintainer-guide/governance/                    /docs/latest/contribute/governance/ 301!
```
```suggestion
/docs/latest/maintainer-guide/issues                         /docs/latest/maintain/manage-issues 301!
/docs/latest/maintainer-guide/pullrequests                   /docs/latest/maintain/review-pull-requests 301!
/docs/latest/maintainer-guide/releases                       /docs/latest/maintain/manage-releases 301!
/docs/latest/maintainer-guide/governance                     /docs/latest/contribute/governance 301!
```

bpmutter commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/user-guide/integrations/                        /docs/latest/use/integrations/ 301!
+/docs/latest/user-guide/migrating-to-8.0.0/                  /docs/latest/use/migrate-to-8.0.0/ 301!
+/docs/latest/developer-guide/architecture/                   /docs/latest/contribute/architecture/ 301!
+/docs/latest/developer-guide/source-code/                    /docs/latest/contribute/development-environment/ 301!
```
yes, i think so. 

i think an incrementalist approach is safer, more well scoped, and easier to review, even if it means more PRs.

mdjermanovic commented:
For the start, we could add these redirects only for the English site. Once zh-hans.eslint.org is ready for this change, we can remove the condition.

mdjermanovic commented:
`/docs/latest/developer-guide/code-conventions` should be redirected to `/docs/latest/contribute/code-conventions`.

mdjermanovic commented:
`/docs/latest/developer-guide/package-json-conventions` should be redirected to `/docs/latest/contribute/package-json-conventions`

mdjermanovic commented:
`/docs/latest/developer-guide/working-with-rules-deprecated` should be redirected to `/docs/latest/extend/custom-rules-deprecated`.

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,47 @@ eleventyExcludeFromCollections: true
+/docs/latest/maintainer-guide/pullrequests                   /docs/latest/maintain/review-pull-requests 301!
+/docs/latest/maintainer-guide/releases                       /docs/latest/maintain/manage-releases 301!
+/docs/latest/maintainer-guide/governance                     /docs/latest/contribute/governance 301!
+/docs/latest/user-guide/configuring/*                        /docs/latest/use/contribute/:splat 301!
```
```suggestion
/docs/latest/user-guide/configuring/*                        /docs/latest/use/configure/:splat 301!
```

mdjermanovic commented:
If we'd like to enable these redirects only for the English site (`eslint.org`), we can wrap them in `{% if site.language.code == "en" %}`...`{% endif %}`.

bpmutter commented:
> If we'd like to enable these redirects only for the English site (`eslint.org`), we can wrap them in `{% if site.language.code == "en" %}`...`{% endif %}`.

@mdjermanovic would you mind putting suggestions for this change on the PR? 

i don't fully understand what's going on with conditionals statements for the redirects

bpmutter commented:
@mdjermanovic thank you again for a thorough review. added the above suggested redirects and also went through the all file moves in the https://github.com/eslint/eslint/pull/16665 and added some redirects that were covered by the splat to be explicit in [f5e151a](https://github.com/eslint/eslint.org/pull/388/commits/f5e151a5d0cedd23ddf7986d599802f2a940252e)

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,54 @@ eleventyExcludeFromCollections: true
 /docs/maintainer-guide/*            /docs/latest/maintainer-guide/:splat 301!
 /docs/developer-guide/*             /docs/latest/developer-guide/:splat 301!
 
+# Redirects for the ESLint IA Refactor (https://github.com/eslint/rfcs/pull/97)
```
```suggestion
{% if site.language.code == "en" %}

# Redirects for the ESLint IA Refactor (https://github.com/eslint/rfcs/pull/97)
```

mdjermanovic commented on src/static/redirects.njk:
```diff
@@ -31,6 +31,54 @@ eleventyExcludeFromCollections: true
+/docs/latest/user-guide/*                                    /docs/latest/use/:splat 301!
+/docs/latest/developer-guide/*                               /docs/latest/extend/:splat 301!
+/docs/latest/maintainer-guide/*                              /docs/latest/maintain/:splat 301!
+
```
```suggestion

{% endif %}

```

mdjermanovic reviewed (commented):
> @mdjermanovic would you mind putting suggestions for this change on the PR

Here it is

nzakas commented:
If we are doing this just for the English site now, then maybe we should put these into the English site file instead?
https://github.com/eslint/eslint.org/blob/main/src/_data/sites/en.yml#L23-L29

I've been trying to avoid checking the language of the site inside of other files.

mdjermanovic commented:
> If we are doing this just for the English site now, then maybe we should put these into the English site file instead? https://github.com/eslint/eslint.org/blob/main/src/_data/sites/en.yml#L23-L29
> 
> I've been trying to avoid checking the language of the site inside of other files.

I think it would be easier to leave this here, as it's only temporary. Once we update the `zh-hans` docs site too, we can just remove the condition.

mdjermanovic reviewed (approved):
LGTM, thanks!

This should be merged right after the release that includes https://github.com/eslint/eslint/pull/16665
